### PR TITLE
First Phase of Tracing Transition

### PR DESCRIPTION
**Pull request by cwfitzgerald**
## Connections

First step in the implementation of #491. https://github.com/gfx-rs/wgpu-rs/pull/395

## Description

This adds the tracing crate, implements a tracing "layer" for chrome tracing, and instruments every entrypoint into wgpu.

Tracing is added as a main dependency. A feature is added called `subscriber` which guards the tracing and default logger implementation, as that adds 3 dependencies. 

The main macro is there to make creating a span a simple one line process. This macro will come in useful in the next couple stages. Use of this macro is used unqualified with it imported into scope as that style allows IntelliJ ides to actually find the macro.

I also removed a really annoying warning that was driving me crazy.

This PR does not make sure the logging output from tracing is up to snuff, that will be done when logging output and conversion is the priority.

Both commits should compile individually, so shouldn't need to be squashed.

## Testing

This PR was tested with the wgpu-rs PR on various examples, as well as my personal project.

### CODE REVIEW

**Comment by cwfitzgerald** [comment]
Rebase is finished, all commits still individually compile, so this should be all ready to review.

File: wgpu-core/src/lib.rs [hunk 1]
```diff
@@ -31,6 +31,7 @@ pub mod device;
 pub mod hub;
 pub mod id;
 pub mod instance;
+pub mod logging;
```
Comment (kvark) [inline wgpu-core/src/lib.rs:34]: None of the other modules we have ends with "ing". Could this one be just `log`?

File: wgpu-core/src/logging/subscriber.rs [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+///   at that location.  
+pub fn initialize_default_subscriber(chrome_trace_path: Option<impl AsRef<Path>>) {
+    let chrome_tracing_layer_opt =
+        chrome_trace_path.map(|path| ChromeTracingLayer::with_file(path).unwrap());
+
+    // Tracing currently doesn't support type erasure with layer composition
+    if let Some(chrome_tracing_layer) = chrome_tracing_layer_opt {
+        tracing::subscriber::set_global_default(
+            tracing_subscriber::Registry::default()
+                .with(chrome_tracing_layer)
```
Comment (kvark) [inline wgpu-core/src/logging/subscriber.rs:36]: if this is the only line that's different, can we move out the whole thing from the if/else arms?

File: wgpu-core/src/logging/subscriber.rs [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+
+        let mut file = self.file.lock().unwrap();
+        writeln!(
+            file,
+            r#"{{ "name": "{}", "cat": "{}", "ph": "{}", "ts": {}, "pid": {}, "tid": {} {} }},"#,
+            name,
+            category,
+            event_type_str,
+            diff_in_us,
+            *CURRENT_PROCESS_ID,
```
Comment (kvark) [inline wgpu-core/src/logging/subscriber.rs:147]: same for the process ID - not clear to me if it's worth caching in a static?

File: wgpu-core/src/logging/subscriber.rs [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+        )
+        .unwrap();
+    }
+}
+
+static CURRENT_PROCESS_ID: once_cell::sync::Lazy<u32> =
+    once_cell::sync::Lazy::new(|| std::process::id());
+
+thread_local! {
+    static CURRENT_THREAD_ID: usize = thread_id::get();
```
[the remark this answers is not in the document]
Comment (cwfitzgerald) [reply wgpu-core/src/logging/subscriber.rs:55]: Well... There's two problems with `std::thread::current().id()`. First is that it isn't necessarily the underlying thread id on the system. The second is that it's an opaque handle that is only debug printable. If I debug print it, that gives me `ThreadId(12931092)` which ofc, isn't a valid json value. There is an api to get the underlying NonZeroUsize out of a thread id..... but it's nightly. Bleh.

File: wgpu-core/src/command/bundle.rs [hunk 5]
```diff
@@ -898,6 +911,7 @@ pub mod bundle_ffi {
         _bundle: &mut RenderBundleEncoder,
         _label: RawString,
     ) {
+        span!(_guard, ERROR, "RenderBundle::insert_debug_marker");
```
[the remark this answers is not in the document]
Comment (cwfitzgerald) [reply wgpu-core/src/command/bundle.rs:914]: The error level of the spans dictates what logging level they will display at with the built in formatter. We want the entry points to always show, so we want it at ERROR (and below).

File: wgpu-core/src/logging/subscriber.rs [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+        let mut file = self.file.lock().unwrap();
+        writeln!(
+            file,
+            r#"{{ "name": "{}", "cat": "{}", "ph": "{}", "ts": {}, "pid": {}, "tid": {} {} }},"#,
+            name,
+            category,
+            event_type_str,
+            diff_in_us,
+            *CURRENT_PROCESS_ID,
+            CURRENT_THREAD_ID.with(|v| *v),
```
[the remark this answers is not in the document]
Comment (cwfitzgerald) [reply wgpu-core/src/logging/subscriber.rs:148]: Iirc, on x64 the pointer to TLS is stored in... one (`fs` maybe) of the two fancy registers that you can do `fs + [compound addressing]` with (similarly to how you can do rip relative addressing). This pointer is setup by the os/libc automagically.

File: wgpu-core/src/lib.rs (hunk 1 above)
Comment (cwfitzgerald) [reply wgpu-core/src/lib.rs:34]: I named it such to avoid conflict with `log`.

File: wgpu-core/src/logging/subscriber.rs (hunk 2 above)
Comment (cwfitzgerald) [reply wgpu-core/src/logging/subscriber.rs:36]: They have different types, `.with` results in `Layered<TheLastThing, ThisThing>` so I have to carry it through all the way to the point of type erasure (set_global_default)

File: wgpu-core/src/logging/subscriber.rs (hunk 3 above)
Comment (cwfitzgerald) [reply wgpu-core/src/logging/subscriber.rs:147]: I expect the actual os call to be significant enough that's it's worth caching, even if it's just a context switch in hot code.

Comment (cwfitzgerald) [reply wgpu-core/src/logging/subscriber.rs:147]: Oh... yes of course :)

File: wgpu-core/src/lib.rs (hunk 1 above)
Comment (cwfitzgerald) [reply wgpu-core/src/lib.rs:34]: Yup!

File: wgpu-core/src/command/bundle.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (cwfitzgerald) [reply wgpu-core/src/command/bundle.rs:914]: Discussed on the matrix to follow the current logging levels.

**Comment by cwfitzgerald** [comment]
Alright, everything should be up to date here.

File: wgpu-core/src/logging/subscriber.rs [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+        let mut file = self.file.lock().unwrap();
+        writeln!(
+            file,
+            r#"{{ "name": "{}", "cat": "{}", "ph": "{}", "ts": {}, "pid": {}, "tid": {} {} }},"#,
+            name,
+            category,
+            event_type_str,
+            diff_in_us,
+            self.process_id,
+            CURRENT_THREAD_ID.with(|v| *v),
```
Comment (kvark) [inline wgpu-core/src/logging/subscriber.rs:147]: Can we just do `thread_id::get()` here if we don't expect this to be used from the same thread?

Comment (cwfitzgerald) [reply wgpu-core/src/logging/subscriber.rs:147]: Discussed in matrix, leaving as is.

**Comment by cwfitzgerald** [comment]
bors r=kvark

**Pull request merged**